protocol/nats_jetstream/v3/receiver.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// Receive implements Receiver.Receive.
+func (r *receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+type consumer struct {
+	receiver
+
+	conn                  *nats.Conn
+	jetSteamOpts          []jetstream.JetStreamOpt
+	jetStream             jetstream.JetStream
+	consumerConfig        *jetstream.ConsumerConfig
+	orderedConsumerConfig *jetstream.OrderedConsumerConfig
+	pullConsumeOpts       []jetstream.PullConsumeOpt
+	jetstreamConsumer     jetstream.Consumer
+
+	subMtx        sync.Mutex
+	internalClose chan struct{}
+	connOwned     bool
+}
+
+// NewConsumer consumes from filterSubjects given in ConsumerConfig or OrderedConsumerConfig
+// embedded in ConsumerOption.
+// See: WithConsumerConfig(...) and WithOrderedConsumerConfig(...)
+func NewConsumer(ctx context.Context, url string, natsOpts []nats.Option, opts ...ProtocolOption) (*consumer, error) {
+	conn, err := nats.Connect(url, natsOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := NewConsumerFromConn(ctx, conn, opts...)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	c.connOwned = true
+
+	return c, err
+}
+
+// NewConsumerFromConn consumes from filterSubjects given in ConsumerConfig or OrderedConsumerConfig
+// embedded in ConsumerOption. See: WithConsumerConfig(...) and WithOrderedConsumerConfig(...)
+func NewConsumerFromConn(ctx context.Context, conn *nats.Conn, opts ...ProtocolOption) (*consumer, error) {
+	c := &consumer{
+		receiver:      *newReceiver(),
+		conn:          conn,
+		internalClose: make(chan struct{}, 1),
+	}
+
+	if err := c.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	var err error
+	if c.jetStream, err = jetstream.New(conn, c.jetSteamOpts...); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (c *consumer) OpenInbound(ctx context.Context) error {
+	c.subMtx.Lock()
+	defer c.subMtx.Unlock()
+
+	var consumeContext jetstream.ConsumeContext
+	var err error
+	if err = c.createConsumer(ctx); err != nil {
+		return err
+	}
+	if consumeContext, err = c.jetstreamConsumer.Consume(c.MsgHandler, c.pullConsumeOpts...); err != nil {
+		return err
+	}
+
+	// Wait until external or internal context done
+	select {
+	case <-ctx.Done():
+	case <-c.internalClose:
+	}
+
+	// Finish to consume messages in the queue and close the subscription
+	if consumeContext != nil {
+		consumeContext.Drain()
+	}
+	return nil
+}
+
+// Close implements Closer.Close
+func (c *consumer) Close(ctx context.Context) error {
+	// Before closing, let's be sure OpenInbound completes
+	// We send a signal to close and then we lock on subMtx in order
+	// to wait OpenInbound to finish draining the queue
+	c.internalClose <- struct{}{}
+	c.subMtx.Lock()
+	defer c.subMtx.Unlock()
+
+	if c.connOwned {
+		c.conn.Close()
```

Comment:
nit: is there a possibility of encountering a panic when calling `c.conn.Close()` more than once?

Reply:
added a nil check